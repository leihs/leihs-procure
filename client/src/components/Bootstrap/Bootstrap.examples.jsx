import React from 'react'

import {
  Row,
  Col,
  Button,
  Badge,
  FormGroup,
  FormField,
  InputDate,
  ButtonRadio
} from './index'

const _space_ = ' '

export const examples = [
  {
    name: 'Button',
    content: (
      <React.Fragment>
        <Button>Button</Button>
        {_space_}
        <Button color="primary" title="test title">
          Button with title
        </Button>
        {_space_}
        <Button color="primary">Button primary</Button>
        {_space_}
        <Button color="secondary">Button secondary</Button>
        {_space_}
        <Button color="success">Button success</Button>
        {_space_}
        <Button color="danger">Button danger</Button>
        {_space_}
        <Button color="warning">Button warning</Button>
        {_space_}
        <Button color="info">Button info</Button>
        {_space_}
        <Button color="light">Button light</Button>
        {_space_}
        <Button color="dark">Button dark</Button>
        {_space_}
        {_space_}
        <Button onClick={e => window.alert('clicked!')}>Button</Button>
        {_space_}
        <Button color="link" title="test title" className="my-class">
          Button
        </Button>
      </React.Fragment>
    )
  },
  {
    name: 'Badge',
    content: (
      <React.Fragment>
        <h3>
          Heading <Badge>Badge</Badge>
        </h3>
        <Badge>Badge</Badge>
        {_space_}
        <Badge primary>Badge primary</Badge>
        {_space_}
        <Badge secondary>Badge secondary</Badge>
        {_space_}
        <Badge success>Badge success</Badge>
        {_space_}
        <Badge danger>Badge danger</Badge>
        {_space_}
        <Badge warning>Badge warning</Badge>
        {_space_}
        <Badge info>Badge info</Badge>
        {_space_}
        <Badge light>Badge light</Badge>
        {_space_}
        <Badge dark>Badge dark</Badge>
      </React.Fragment>
    )
  },
  {
    name: 'FormField',
    content: (() => {
      class DemoForm extends React.Component {
        state = {
          a: 'some text',
          b: 'some static text',
          c: 'some text',
          d: '23.42',
          e: '42',
          f: true,
          g: null,
          h: null
        }
        render() {
          return (
            <Row>
              <Col>
                <FormField
                  type="text"
                  label="type=text"
                  name="a"
                  required
                  value={this.state.a}
                  onChange={e => this.setState({ a: e.target.value })}
                />
                <FormField
                  name="b"
                  type="text-static"
                  label="type=text-static"
                  required
                  value={this.state.b}
                  onChange={e => this.setState({ b: e.target.value })}
                />
                <FormField
                  name="c"
                  type="textarea"
                  label="type=textarea"
                  required
                  value={this.state.c}
                  onChange={e => this.setState({ c: e.target.value })}
                />
              </Col>
              <Col>
                <FormField
                  name="d"
                  type="number"
                  label="type=number"
                  required
                  value={this.state.d}
                  onChange={e => this.setState({ d: e.target.value })}
                />
                <FormField
                  name="e"
                  type="number-integer"
                  label="type=number-integer"
                  required
                  value={this.state.e}
                  onChange={e => this.setState({ e: e.target.value })}
                />
                <FormField
                  name="f"
                  id="example_formfield_f"
                  type="checkbox"
                  label="type=checkbox"
                  required
                  checked={this.state.f}
                  onChange={e => this.setState({ f: e.target.value })}
                />
                <FormField
                  name="g"
                  id="example_formfield_g"
                  type="checkbox"
                  label="type=checkbox"
                  required
                  checked={this.state.g}
                  onChange={e => this.setState({ g: e.target.value })}
                />
                <hr />
                <FormGroup label="ButtonRadio Radio Button">
                  <ButtonRadio
                    name="h"
                    id="example_formfield_h"
                    label="<ButtonRadio/>"
                    required
                    value={this.state.h}
                    onChange={e => this.setState({ h: e.target.value })}
                    options={[
                      { value: 'one', label: 'Eins' },
                      { value: 'two', label: 'Zwei' }
                    ]}
                  />
                </FormGroup>
              </Col>
              <Col>
                <pre>{JSON.stringify(this.state, 0, 2)}</pre>
              </Col>
            </Row>
          )
        }
      }
      return (
        <React.Fragment>
          <h2>controlled</h2>
          <DemoForm />
          <h2>controlled, with validation styles</h2>
          <div className="was-validated">
            <DemoForm />
          </div>
        </React.Fragment>
      )
    })()
  }
  // {
  //   name: 'InputDate',
  //   content: (
  //     <React.Fragment>
  //       no value
  //       <br />
  //       <InputDate />
  //       <hr />
  //       with value <code>1985-10-26T08:15:00.000Z</code>
  //       <br />
  //       <InputDate value="1985-10-26T08:15:00.000Z" />
  //     </React.Fragment>
  //   )
  // }
]
