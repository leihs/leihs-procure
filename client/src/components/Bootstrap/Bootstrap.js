import React, { Fragment as F } from 'react'
import PropTypes from 'prop-types'
import cx from 'classnames/dedupe'
import f from 'lodash'

import BsButton from 'reactstrap/lib/Button'
import BsButtonGroup from 'reactstrap/lib/ButtonGroup'

import Icon from '../Icons'
import { StatefulInput } from './StatefulForm'

const BOOTSTRAP_BREAKPOINTS = ['sm', 'md', 'lg', 'xl']
const BOOTSTRAP_MODIFIERS = [
  'primary',
  'secondary',
  'success',
  'danger',
  'warning',
  'info',
  'light',
  'dark'
]

// https://getbootstrap.com/docs/4.0/utilities/spacing/#notation
const bsSizeUtils = ({ m, p, ...props }) => {
  const { cls, className, ...restProps } = props
  function extract(letter, prop) {
    if (f.isArray(prop)) return prop.map(p => extract(letter, p))
    if (!f.isString(prop)) return
    let parts = prop.split(' ')
    if (parts.length > 1) return extract(letter, parts)
    return `${letter}${prop}`
  }
  const margin = extract('m', m)
  const padding = extract('p', p)
  return [cx(margin, padding, cls, className), restProps]
}

const Node = ({ tag = 'div', ...props }) => {
  // eslint-disable-next-line no-debugger
  const Tag = tag
  const [bsClasses, restProps] = bsSizeUtils(props)
  return <Tag {...restProps} className={bsClasses} />
}
Node.propTypes = {
  tag: PropTypes.oneOfType([PropTypes.string, PropTypes.func])
}

export { Node as __Node } // "internal" usage only

export const Div = props => Node(props)
export const Pre = props => Node({ ...props, tag: 'span' })
export const Span = props => Node({ ...props, tag: 'span' })
export const Code = props => Node({ ...props, tag: 'span' })
export const Hr = props => Node({ ...props, tag: 'span' })
export const Main = props => Node({ ...props, tag: 'main' })
export const Label = props => Node({ ...props, tag: 'label' })
export const Small = props => Node({ ...props, tag: 'small' })

export const Container = ({ fluid, ...props }) =>
  Node({ ...props, cls: [fluid ? 'container-fluid' : 'container', props.cls] })

export const Row = ({ form, ...props }) =>
  Node({ ...props, cls: [form ? 'form-row' : 'row', props.cls] })

export const Col = ({ order, cls, ...props }) => {
  const restProps = f.omit(props, BOOTSTRAP_BREAKPOINTS)
  const breakpoint = f.first(
    f.intersection(f.keys(props), BOOTSTRAP_BREAKPOINTS)
  )
  const breakpointVal = props[breakpoint]
  const colCls = breakpoint
    ? f.isString(breakpointVal) || f.isNumber(breakpointVal)
      ? `col-${breakpoint}-${breakpointVal}`
      : `col-${breakpoint}`
    : 'col'
  const orderCls =
    order && (breakpoint ? `order-${breakpoint}-${order}` : `order-${order}`)
  return Node({ ...restProps, cls: cx(colCls, orderCls, cls) })
}

export const Anchor = ({ target, rel, ...p }) => {
  if (!rel && target === '_blank') rel = 'noopener noreferrer'
  return <Node tag="a" target={target} rel={rel} {...p} />
}

export const Button = ({ massive, className, cls, ...props }) => (
  <BsButton
    type="button" // default in case not given, otherwise its 'submit'
    {...props}
    className={cx(className, cls, { 'btn-massive': massive })}
  />
)
Button.propTypes = {
  className: PropTypes.string,
  massive: PropTypes.bool
}

export { BsButtonGroup as ButtonGroup }

export const Badge = props => {
  const restProps = f.omit(props, BOOTSTRAP_MODIFIERS)
  const mod =
    f.first(f.intersection(f.keys(props), BOOTSTRAP_MODIFIERS)) ||
    BOOTSTRAP_MODIFIERS[0]
  return <Span {...restProps} cls={[props.cls, 'badge', `badge-${mod}`]} />
}
Badge.propTypes = {
  cls: PropTypes.any // todo: classnames.proptypes
}

export const FormGroup = ({
  id,
  label,
  hideLabel,
  labelSmall,
  labelWidth = '4',
  helpText,
  children,
  horizontal,
  ...props
}) => {
  const cls = [
    { 'form-row': horizontal },
    props.cls,
    props.className,
    'form-group'
  ]

  const inputWrapCls = horizontal && 'col-sm'

  const labelContent = !!(label || labelSmall) && (
    <F>
      {label}
      {!!labelSmall && (
        <F>
          {' '}
          <Small>{labelSmall}</Small>
        </F>
      )}
    </F>
  )

  return (
    <Node {...props} cls={cls}>
      {labelContent && (
        <Label
          htmlFor={id}
          cls={{
            'sr-only': hideLabel,
            [`col-form-label col-sm-${labelWidth}`]: horizontal
          }}
        >
          {labelContent}{' '}
        </Label>
      )}
      {!!children && <Div cls={inputWrapCls}>{children} </Div>}
      {helpText && (
        <Small id={`${id}--Help`} cls="form-text text-muted">
          {f.trim(helpText)}{' '}
        </Small>
      )}
    </Node>
  )
}
FormGroup.propTypes = {
  id: PropTypes.string,
  label: PropTypes.node,
  hideLabel: PropTypes.bool,
  labelSmall: PropTypes.node,
  helpText: PropTypes.node,
  children: PropTypes.node
}

export const InputText = props => (
  <StatefulInput {...props}>
    {inputProps => (
      <Node
        {...props}
        {...inputProps}
        tag="input"
        type={props.type || 'text'}
        cls={['form-control', inputProps.cls]}
        value={inputProps.value || props.value}
        autoComplete={
          props.autoComplete === true // let the browser decide if true, or
            ? null // pass through given prop or turn off if falsy/nothing
            : props.autoComplete || 'off'
        }
      />
    )}
  </StatefulInput>
)

const InputFieldPropTypes = {
  id: PropTypes.string,
  children: PropTypes.oneOf([null, undefined]),
  label: PropTypes.node,
  name: PropTypes.string.isRequired,
  placeholder: PropTypes.string,
  type: PropTypes.string, // enum, already checked at runtime
  hidden: PropTypes.bool,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.bool
  ])
}
export const InputField = ({
  id,
  children,
  inputLabel,
  name,
  placeholder,
  type = 'text',
  value,
  label,
  autoComplete,
  minRows,
  maxRows,
  className,
  hidden,
  ...rest
}) => {
  // FIXME: better hiding like html's input.hidden?
  if (hidden) return false

  const supportedTypes = [
    'text',
    'text-static',
    'textarea',
    'number',
    'number-integer',
    'checkbox'
  ]
  if (type === 'checkbox' && !id) {
    throw new Error('Input `type=checkbox` is missing required `id`!')
  }

  if (!f.includes(supportedTypes, type)) {
    throw new Error('Unsupported Input Type!')
  }

  let tag = 'input'
  let mainClass = 'form-control'
  let inputNode
  if (!name) name = id
  let inputProps = {
    className: cx(mainClass, className),
    ...rest
  }

  if (type === 'text') {
    tag = InputText
  }

  if (type === 'text-static') {
    tag = inputProps.tag || 'span'
    inputNode = (
      <Node
        tag={tag}
        id={id}
        {...f.omit(inputProps, 'onChange', 'aria-describedby')}
        className={cx(inputProps.className, 'form-control-plaintext', {
          'form-control': false
        })}
      >
        {value}
      </Node>
    )
  }

  if (type === 'textarea') {
    tag = type
    type = null
    minRows = f.presence(minRows) || 3
    maxRows = f.presence(maxRows) || 7
    inputProps = {
      rows: Math.max(
        minRows,
        Math.min(maxRows, String(value).split('\n').length)
      ),
      ...inputProps
    }
  }

  if (type === 'number-integer') {
    type = 'number'
    inputProps = { min: 0, step: 1, ...inputProps }
  }

  if (type === 'checkbox') {
    inputProps = {
      ...inputProps,
      className: cx(inputProps.className, 'custom-control-input', {
        'form-control': false
      }),
      checked: value
    }
    value = null
  }

  const inputAutoComplete =
    autoComplete === true // let the browser decide if true, or
      ? null // pass through given prop or turn off if falsy/nothing
      : autoComplete || 'off'

  inputNode = inputNode || (
    <Node
      {...inputProps}
      tag={tag}
      type={type}
      id={id}
      name={name}
      value={value === null ? '' : value}
      autoComplete={inputAutoComplete}
      placeholder={placeholder}
    />
  )

  if (type === 'checkbox') {
    inputNode = (
      <div className="custom-control custom-checkbox">
        {inputNode}
        <label className="custom-control-label" htmlFor={id}>
          {label}
        </label>
      </div>
    )
  }

  return inputNode
}

InputField.propTypes = InputFieldPropTypes

const FormFieldPropTypes = {
  beforeInput: PropTypes.node,
  afterInput: PropTypes.node,
  helpText: PropTypes.node,
  invalidFeedback: PropTypes.string,
  id: PropTypes.string,
  children: PropTypes.oneOf([null, undefined]),
  hideLabel: PropTypes.bool,
  labelSmall: PropTypes.node,
  name: PropTypes.string.isRequired,
  placeholder: PropTypes.string,
  type: PropTypes.string, // enum, already checked at runtime
  hidden: PropTypes.bool,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.bool
  ])
}
export const FormField = ({
  beforeInput,
  afterInput,
  invalidFeedback,
  id,
  helpText,
  children,
  label,
  hideLabel,
  labelSmall,
  inputLabel,
  horizontal,
  ...restProps
}) => {
  if (children) {
    throw new Error('`children` not supported! Use `FormGroup` or instead.')
  }

  const inputNode = (
    <InputField
      id={id}
      label={inputLabel}
      aria-describedby={helpText ? `${id}--Help` : null}
      {...restProps}
    />
  )
  return (
    <FormGroup
      label={label}
      hideLabel={hideLabel}
      labelSmall={labelSmall}
      helpText={helpText}
      horizontal={horizontal}
    >
      {beforeInput}
      {inputNode}
      {!!invalidFeedback && (
        <div className="invalid-feedback">{invalidFeedback}</div>
      )}
      {afterInput}
    </FormGroup>
  )
}
FormField.propTypes = FormFieldPropTypes

// like bootstrap docs
// export const FilePicker = ({ id, name }) => (
//   <div className="custom-file">
//     <input type="file" className="custom-file-input" id={id} name={name} />
//     <label className="custom-file-label" htmlFor="customFile">
//       <Icon.Paperclip /> {t('form_filepicker_label')}
//     </label>
//   </div>
// )

export const FilePicker = ({ id, name, label, onChange, ...props }) => (
  <F>
    <label className="mb-0 btn btn-sm btn-block btn-outline-secondary text-left">
      <Icon.Paperclip /> {label}
      <input
        id={id}
        name={name}
        type="file"
        style={{ opacity: 0, width: 0 }}
        onChange={onChange}
        {...props}
      />
    </label>
  </F>
)
FilePicker.propTypes = {
  id: PropTypes.string,
  name: PropTypes.string,
  label: PropTypes.node
}
