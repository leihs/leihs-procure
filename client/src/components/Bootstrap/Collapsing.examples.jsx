import React from 'react'
import cx from 'classnames'
// import f from 'lodash'

import { Row, Col, Collapsing } from '.'

export const examples = [
  {
    title: 'basic',
    content: (
      <Row>
        <Col sm>
          <Collapsing id="collapse-example-basic" startOpen>
            {({ isOpen, togglerProps, collapsedProps, Caret }) => (
              <div>
                <div {...togglerProps}>
                  <Caret spaced />
                  Collapsed Title
                </div>
                {isOpen && <div {...collapsedProps}>Collapsed Content!</div>}
              </div>
            )}
          </Collapsing>
        </Col>
        <Col sm>
          <pre>
            <code>{`
<Collapsing id="collapse-example-1" startOpen>
  {({ isOpen, togglerProps, collapsedProps, Caret }) => (
    <div>
      <div {...togglerProps}>
        <Caret spaced />
        Collapsed Title
      </div>
      {isOpen && <div {...collapsedProps}>Collapsed Content!</div>}
    </div>
  )}
</Collapsing>
        `}</code>
          </pre>
        </Col>
      </Row>
    )
  },
  {
    title: 'card style',
    content: (
      <Row>
        <Col sm>
          <Collapsing id="collapse-example-card" startOpen>
            {({ isOpen, togglerProps, collapsedProps, Caret }) => (
              <div className={cx('card mb-3')}>
                <div
                  className={cx('card-header cursor-pointer pl-2', {
                    'border-bottom-0': !isOpen
                  })}
                  {...togglerProps}
                >
                  <h1 className="mb-0 h4 d-inline-block">
                    <Caret spaced />
                    Collapsed Title
                  </h1>
                </div>

                {isOpen && (
                  <div className={cx('card-body pl-2')} {...collapsedProps}>
                    <p className="h5 p-4">Collapsed Content!</p>
                  </div>
                )}
              </div>
            )}
          </Collapsing>
        </Col>
      </Row>
    )
  }
]
