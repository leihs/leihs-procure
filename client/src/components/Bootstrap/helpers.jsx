import f from 'lodash'

// build `options` prop for a `<Select/>`, e.g.
// `{id: 23, name: 'Foo'}` to `{value: 23, label: 'Foo'}`
export const optionFromObject = (
  obj,
  path,
  valueKey = 'id',
  labelKey = 'name'
) => {
  const item = path ? f.get(obj.path) : obj
  return { label: f.get(item, labelKey), value: f.get(item, valueKey) }
}

// apply 'optionFromObject' to a list of objects
export const optionsFromList = (list, ...args) =>
  f.map(list, item => optionFromObject(item, ...args))
