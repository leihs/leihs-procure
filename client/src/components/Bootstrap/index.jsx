export * from './Bootstrap'
// export const __Node = null // don't export this into rest of app

export * from './helpers'

export * from './Navs'
export { Collapsing } from './Collapsing'
export { default as Tooltipped } from './Tooltipped'

export { default as StatefulForm, StatefulInput } from './StatefulForm'

export { default as InputDate } from './InputDate'
export { default as Select } from './Select'
export { default as ButtonRadio } from './ButtonRadio'
export { default as MultiSelect } from './MultiSelect'

export { default as InputFileUpload } from './InputFileUpload'

export {
  Alert,
  UncontrolledAlert as AlertDismissable,
  ListGroup,
  ListGroupItem,
  ButtonToolbar
} from 'reactstrap'
