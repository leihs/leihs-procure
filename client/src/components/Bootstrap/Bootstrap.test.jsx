// import React from 'react'
import renderer from 'react-test-renderer'
import { examples } from './Bootstrap.examples'

examples.forEach(({ name, content }) => {
  it(`${name}: renders correctly`, () => {
    expect(renderer.create(content).toJSON()).toMatchSnapshot()
  })
})
