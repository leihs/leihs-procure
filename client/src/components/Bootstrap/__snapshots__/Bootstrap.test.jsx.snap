// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Badge: renders correctly 1`] = `
[
  <h3>
    Heading 
    <span
      className="badge badge-primary"
    >
      Badge
    </span>
  </h3>,
  <span
    className="badge badge-primary"
  >
    Badge
  </span>,
  " ",
  <span
    className="badge badge-primary"
  >
    Badge primary
  </span>,
  " ",
  <span
    className="badge badge-secondary"
  >
    Badge secondary
  </span>,
  " ",
  <span
    className="badge badge-success"
  >
    Badge success
  </span>,
  " ",
  <span
    className="badge badge-danger"
  >
    Badge danger
  </span>,
  " ",
  <span
    className="badge badge-warning"
  >
    Badge warning
  </span>,
  " ",
  <span
    className="badge badge-info"
  >
    Badge info
  </span>,
  " ",
  <span
    className="badge badge-light"
  >
    Badge light
  </span>,
  " ",
  <span
    className="badge badge-dark"
  >
    Badge dark
  </span>,
]
`;

exports[`Button: renders correctly 1`] = `
[
  <button
    aria-label={null}
    className="btn btn-secondary"
    onClick={[Function]}
    type="button"
  >
    Button
  </button>,
  " ",
  <button
    aria-label={null}
    className="btn btn-primary"
    onClick={[Function]}
    title="test title"
    type="button"
  >
    Button with title
  </button>,
  " ",
  <button
    aria-label={null}
    className="btn btn-primary"
    onClick={[Function]}
    type="button"
  >
    Button primary
  </button>,
  " ",
  <button
    aria-label={null}
    className="btn btn-secondary"
    onClick={[Function]}
    type="button"
  >
    Button secondary
  </button>,
  " ",
  <button
    aria-label={null}
    className="btn btn-success"
    onClick={[Function]}
    type="button"
  >
    Button success
  </button>,
  " ",
  <button
    aria-label={null}
    className="btn btn-danger"
    onClick={[Function]}
    type="button"
  >
    Button danger
  </button>,
  " ",
  <button
    aria-label={null}
    className="btn btn-warning"
    onClick={[Function]}
    type="button"
  >
    Button warning
  </button>,
  " ",
  <button
    aria-label={null}
    className="btn btn-info"
    onClick={[Function]}
    type="button"
  >
    Button info
  </button>,
  " ",
  <button
    aria-label={null}
    className="btn btn-light"
    onClick={[Function]}
    type="button"
  >
    Button light
  </button>,
  " ",
  <button
    aria-label={null}
    className="btn btn-dark"
    onClick={[Function]}
    type="button"
  >
    Button dark
  </button>,
  " ",
  " ",
  <button
    aria-label={null}
    className="btn btn-secondary"
    onClick={[Function]}
    type="button"
  >
    Button
  </button>,
  " ",
  <button
    aria-label={null}
    className="my-class btn btn-link"
    onClick={[Function]}
    title="test title"
    type="button"
  >
    Button
  </button>,
]
`;

exports[`FormField: renders correctly 1`] = `
[
  <h2>
    controlled
  </h2>,
  <div
    className="row"
  >
    <div
      className="col"
    >
      <div
        className="form-group"
      >
        <label
          className=""
        >
          type=text
           
        </label>
        <div
          className=""
        >
          <input
            aria-describedby={null}
            autoComplete="off"
            className="form-control"
            name="a"
            onChange={[Function]}
            required={true}
            type="text"
            value="some text"
          />
           
        </div>
      </div>
      <div
        className="form-group"
      >
        <label
          className=""
        >
          type=text-static
           
        </label>
        <div
          className=""
        >
          <span
            className="form-control-plaintext"
            required={true}
          >
            some static text
          </span>
           
        </div>
      </div>
      <div
        className="form-group"
      >
        <label
          className=""
        >
          type=textarea
           
        </label>
        <div
          className=""
        >
          <textarea
            aria-describedby={null}
            autoComplete="off"
            className="form-control"
            name="c"
            onChange={[Function]}
            required={true}
            rows={3}
            type={null}
            value="some text"
          />
           
        </div>
      </div>
    </div>
    <div
      className="col"
    >
      <div
        className="form-group"
      >
        <label
          className=""
        >
          type=number
           
        </label>
        <div
          className=""
        >
          <input
            aria-describedby={null}
            autoComplete="off"
            className="form-control"
            name="d"
            onChange={[Function]}
            required={true}
            type="number"
            value="23.42"
          />
           
        </div>
      </div>
      <div
        className="form-group"
      >
        <label
          className=""
        >
          type=number-integer
           
        </label>
        <div
          className=""
        >
          <input
            aria-describedby={null}
            autoComplete="off"
            className="form-control"
            min={0}
            name="e"
            onChange={[Function]}
            required={true}
            step={1}
            type="number"
            value="42"
          />
           
        </div>
      </div>
      <div
        className="form-group"
      >
        <label
          className=""
        >
          type=checkbox
           
        </label>
        <div
          className=""
        >
          <div
            className="custom-control custom-checkbox"
          >
            <input
              aria-describedby={null}
              autoComplete="off"
              className="custom-control-input"
              id="example_formfield_f"
              name="f"
              onChange={[Function]}
              required={true}
              type="checkbox"
              value=""
            />
            <label
              className="custom-control-label"
              htmlFor="example_formfield_f"
            />
          </div>
           
        </div>
      </div>
      <div
        className="form-group"
      >
        <label
          className=""
        >
          type=checkbox
           
        </label>
        <div
          className=""
        >
          <div
            className="custom-control custom-checkbox"
          >
            <input
              aria-describedby={null}
              autoComplete="off"
              className="custom-control-input"
              id="example_formfield_g"
              name="g"
              onChange={[Function]}
              required={true}
              type="checkbox"
              value=""
            />
            <label
              className="custom-control-label"
              htmlFor="example_formfield_g"
            />
          </div>
           
        </div>
      </div>
      <hr />
      <div
        className="form-group"
      >
        <label
          className=""
        >
          ButtonRadio Radio Button
           
        </label>
        <div
          className=""
        >
          <div
            className="ui-btn-radio btn-group btn-group-toggle d-flex ui-btn-radio-state-invalid"
            label="<ButtonRadio/>"
            role="radiogroup"
          >
            <label
              className="btn btn-block btn-outline-secondary m-0 text-left font-weight-bold"
              htmlFor="example_formfield_h_radio_0"
              tabIndex="-1"
            >
              <input
                aria-checked={false}
                checked={false}
                className="sr-only mt-4 ml-2"
                id="example_formfield_h_radio_0"
                name="h"
                onChange={[Function]}
                onClick={[Function]}
                required={true}
                tabIndex={0}
                type="radio"
                value="one"
              />
              <svg
                aria-hidden="true"
                className="svg-inline--fa fa-circle fa-w-16 mr-2"
                data-icon="circle"
                data-prefix="far"
                focusable="false"
                role="img"
                style={{}}
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm0 448c-110.5 0-200-89.5-200-200S145.5 56 256 56s200 89.5 200 200-89.5 200-200 200z"
                  fill="currentColor"
                  style={{}}
                />
              </svg>
              Eins
            </label>
            <label
              className="btn btn-block btn-outline-secondary m-0 text-left font-weight-bold border-left-0"
              htmlFor="example_formfield_h_radio_1"
              tabIndex="-1"
            >
              <input
                aria-checked={false}
                checked={false}
                className="sr-only mt-4 ml-2"
                id="example_formfield_h_radio_1"
                name="h"
                onChange={[Function]}
                onClick={[Function]}
                required={true}
                tabIndex={-1}
                type="radio"
                value="two"
              />
              <svg
                aria-hidden="true"
                className="svg-inline--fa fa-circle fa-w-16 mr-2"
                data-icon="circle"
                data-prefix="far"
                focusable="false"
                role="img"
                style={{}}
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm0 448c-110.5 0-200-89.5-200-200S145.5 56 256 56s200 89.5 200 200-89.5 200-200 200z"
                  fill="currentColor"
                  style={{}}
                />
              </svg>
              Zwei
            </label>
          </div>
           
        </div>
      </div>
    </div>
    <div
      className="col"
    >
      <pre>
        {
  "a": "some text",
  "b": "some static text",
  "c": "some text",
  "d": "23.42",
  "e": "42",
  "f": true,
  "g": null,
  "h": null
}
      </pre>
    </div>
  </div>,
  <h2>
    controlled, with validation styles
  </h2>,
  <div
    className="was-validated"
  >
    <div
      className="row"
    >
      <div
        className="col"
      >
        <div
          className="form-group"
        >
          <label
            className=""
          >
            type=text
             
          </label>
          <div
            className=""
          >
            <input
              aria-describedby={null}
              autoComplete="off"
              className="form-control"
              name="a"
              onChange={[Function]}
              required={true}
              type="text"
              value="some text"
            />
             
          </div>
        </div>
        <div
          className="form-group"
        >
          <label
            className=""
          >
            type=text-static
             
          </label>
          <div
            className=""
          >
            <span
              className="form-control-plaintext"
              required={true}
            >
              some static text
            </span>
             
          </div>
        </div>
        <div
          className="form-group"
        >
          <label
            className=""
          >
            type=textarea
             
          </label>
          <div
            className=""
          >
            <textarea
              aria-describedby={null}
              autoComplete="off"
              className="form-control"
              name="c"
              onChange={[Function]}
              required={true}
              rows={3}
              type={null}
              value="some text"
            />
             
          </div>
        </div>
      </div>
      <div
        className="col"
      >
        <div
          className="form-group"
        >
          <label
            className=""
          >
            type=number
             
          </label>
          <div
            className=""
          >
            <input
              aria-describedby={null}
              autoComplete="off"
              className="form-control"
              name="d"
              onChange={[Function]}
              required={true}
              type="number"
              value="23.42"
            />
             
          </div>
        </div>
        <div
          className="form-group"
        >
          <label
            className=""
          >
            type=number-integer
             
          </label>
          <div
            className=""
          >
            <input
              aria-describedby={null}
              autoComplete="off"
              className="form-control"
              min={0}
              name="e"
              onChange={[Function]}
              required={true}
              step={1}
              type="number"
              value="42"
            />
             
          </div>
        </div>
        <div
          className="form-group"
        >
          <label
            className=""
          >
            type=checkbox
             
          </label>
          <div
            className=""
          >
            <div
              className="custom-control custom-checkbox"
            >
              <input
                aria-describedby={null}
                autoComplete="off"
                className="custom-control-input"
                id="example_formfield_f"
                name="f"
                onChange={[Function]}
                required={true}
                type="checkbox"
                value=""
              />
              <label
                className="custom-control-label"
                htmlFor="example_formfield_f"
              />
            </div>
             
          </div>
        </div>
        <div
          className="form-group"
        >
          <label
            className=""
          >
            type=checkbox
             
          </label>
          <div
            className=""
          >
            <div
              className="custom-control custom-checkbox"
            >
              <input
                aria-describedby={null}
                autoComplete="off"
                className="custom-control-input"
                id="example_formfield_g"
                name="g"
                onChange={[Function]}
                required={true}
                type="checkbox"
                value=""
              />
              <label
                className="custom-control-label"
                htmlFor="example_formfield_g"
              />
            </div>
             
          </div>
        </div>
        <hr />
        <div
          className="form-group"
        >
          <label
            className=""
          >
            ButtonRadio Radio Button
             
          </label>
          <div
            className=""
          >
            <div
              className="ui-btn-radio btn-group btn-group-toggle d-flex ui-btn-radio-state-invalid"
              label="<ButtonRadio/>"
              role="radiogroup"
            >
              <label
                className="btn btn-block btn-outline-secondary m-0 text-left font-weight-bold"
                htmlFor="example_formfield_h_radio_0"
                tabIndex="-1"
              >
                <input
                  aria-checked={false}
                  checked={false}
                  className="sr-only mt-4 ml-2"
                  id="example_formfield_h_radio_0"
                  name="h"
                  onChange={[Function]}
                  onClick={[Function]}
                  required={true}
                  tabIndex={0}
                  type="radio"
                  value="one"
                />
                <svg
                  aria-hidden="true"
                  className="svg-inline--fa fa-circle fa-w-16 mr-2"
                  data-icon="circle"
                  data-prefix="far"
                  focusable="false"
                  role="img"
                  style={{}}
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm0 448c-110.5 0-200-89.5-200-200S145.5 56 256 56s200 89.5 200 200-89.5 200-200 200z"
                    fill="currentColor"
                    style={{}}
                  />
                </svg>
                Eins
              </label>
              <label
                className="btn btn-block btn-outline-secondary m-0 text-left font-weight-bold border-left-0"
                htmlFor="example_formfield_h_radio_1"
                tabIndex="-1"
              >
                <input
                  aria-checked={false}
                  checked={false}
                  className="sr-only mt-4 ml-2"
                  id="example_formfield_h_radio_1"
                  name="h"
                  onChange={[Function]}
                  onClick={[Function]}
                  required={true}
                  tabIndex={-1}
                  type="radio"
                  value="two"
                />
                <svg
                  aria-hidden="true"
                  className="svg-inline--fa fa-circle fa-w-16 mr-2"
                  data-icon="circle"
                  data-prefix="far"
                  focusable="false"
                  role="img"
                  style={{}}
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm0 448c-110.5 0-200-89.5-200-200S145.5 56 256 56s200 89.5 200 200-89.5 200-200 200z"
                    fill="currentColor"
                    style={{}}
                  />
                </svg>
                Zwei
              </label>
            </div>
             
          </div>
        </div>
      </div>
      <div
        className="col"
      >
        <pre>
          {
  "a": "some text",
  "b": "some static text",
  "c": "some text",
  "d": "23.42",
  "e": "42",
  "f": true,
  "g": null,
  "h": null
}
        </pre>
      </div>
    </div>
  </div>,
]
`;
